File: C++/1405_미친로봇.cpp
#include <iostream>
#include <cstring>
using namespace std;

double p;

void move(int N, int s, int x, int y, bool **visit, double p_dir[], double cur){
	if(s == N){ // N번의 이동이 끝나면 확률 더해줌
		p += cur;
		return;
	}
	
	int dir[4][2] = {{0,1},{0,-1},{1,0},{-1,0}}; // 동서남북
	for(int i = 0; i < 4; i++){ // 동서남북 이동하기 위함
		int nx = x+dir[i][0], ny = y+dir[i][1]; // 이동 후 좌표
		if(nx >= 0 && nx < 2*N+1 && ny >= 0 && ny < 2*N+1){ // 이동 후 좌표가 범위 내에 있고
			if(visit[nx][ny] == false){ // 방문한 적 없으면
				visit[nx][ny] = true;  // 방문 표시 후
				move(N,s+1,nx,ny,visit,p_dir,cur*p_dir[i]); // 재귀
				visit[nx][ny] = false; // 모두 방문 했으면 다음 경우의 수 생각해야 하기 때문에 방문 표시 취소(백트래킹)
			}
		}
	}
}

int main() {
	int N;
	double p_dir[4];
	cin >> N;
	for(int i = 0; i < 4; i++){
		cin >> p_dir[i];
		p_dir[i] /= 100; // 0~100사이의 퍼센트 값을 0~1사이의 확률로 바꾸기 위해 100으로 나눠주기
	}
	
	bool **visit = new bool*[2*N+1];
	for(int i = 0; i < 2*N+1; i++){
		visit[i] = new bool[2*N+1];
		memset(visit[i],false,sizeof(bool)*(2*N+1));
	}
	
	visit[N][N] = true; // 방문 여부 표시
	
	move(N,0,N,N,visit,p_dir,1);
	
	cout << fixed;
	cout.precision(10); // 10^-9 오차까지 허용하기 때문에 10자리까지는 출력해줘야 함
	cout << p;
	
	
	return 0;
}

File: Python/1405_미친로봇.py
p = 0
def moveRobot(N,s,x,y,cur,p_dir,visit):
	if s == N: # N번 모두 이동했다면
		global p 
		p += cur // 현재 확률 전체 확률에 더해줌
		return
    
	dir = [[0,1],[0,-1],[1,0],[-1,0]] # 동서남북
	for i in range(4):
		dx = x + dir[i][0]
		dy = y + dir[i][1] # 이동할 위치
		if dx >= 0 and dx <= 2*N and dy >= 0 and dy <= 2*N: # 범위 내의 위치이고
			if visit[dx][dy] == False: # 방문한 적 없는 위치라면
				visit[dx][dy] = True # 방문 표시
				moveRobot(N,s+1,dx,dy,cur*p_dir[i],p_dir,visit) # 재귀로 다음 step
				visit[dx][dy] = False # 다음 경우의 수 생각하기 위해 방문 표시 제거(현재 dx, dy를 방문한 경우의 수는 모두 확인했음)

if __name__ == '__main__':
	p_dir = list(map(int,input().split()))
	N = p_dir[0]
	p_dir = [x/100 for x in p_dir[1:]]
	
	visit = [[False]*(2*N+1) for _ in range(2*N+1)]
	visit[N][N] = True # 방문 표시(시작 위치)

	moveRobot(N,0,N,N,1,p_dir,visit)
	
	print(p)
